src/libcharon/plugins/kernel_super/kernel_super_ipsec.c

```diff
@@ -3050,34 +3187,62 @@ static bool tunnel_counter_key_compare(const void *key, const void *other_key) {
     const struct tunnel_counter_key *tunnel_1 = (const struct tunnel_counter_key *)key;
     const struct tunnel_counter_key *tunnel_2 = (const struct tunnel_counter_key *)other_key;
 
-	const ike_super_peer_info_t* tunnel_1_peer_info = tunnel_1->super_peer_info;
-	const ike_super_peer_info_t* tunnel_2_peer_info = tunnel_2->super_peer_info;
 
 	return tunnel_1->terminator_id == tunnel_2->terminator_id &&
-			streq(tunnel_1_peer_info->get_tenant_id(tunnel_1_peer_info), tunnel_2_peer_info->get_tenant_id(tunnel_2_peer_info)) &&
-			streq(tunnel_1_peer_info->get_account_id(tunnel_1_peer_info), tunnel_2_peer_info->get_account_id(tunnel_2_peer_info)) &&
-			streq(tunnel_1_peer_info->get_vrf_name(tunnel_1_peer_info), tunnel_2_peer_info->get_vrf_name(tunnel_2_peer_info));
+			streq(tunnel_1->tenant_id, tunnel_2->tenant_id) &&
+			streq(tunnel_1->account_id, tunnel_2->account_id) &&
+			streq(tunnel_1->vrf_name, tunnel_2->vrf_name);
 }
 
 static u_int tunnel_counter_key_hash(const void *key) {
 	const struct tunnel_counter_key *tunnel_key = (const struct tunnel_counter_key*)key;
-	const ike_super_peer_info_t* const super_peer_info = tunnel_key->super_peer_info;
 
 	return chunk_hash_inc(chunk_from_thing(tunnel_key->terminator_id),
-						  chunk_hash_inc(chunk_from_str(super_peer_info->get_tenant_id(super_peer_info)),
-						  chunk_hash_inc(chunk_from_str(super_peer_info->get_account_id(super_peer_info)),
-						  chunk_hash(chunk_from_str(super_peer_info->get_vrf_name(super_peer_info))))));
+						  chunk_hash_inc(chunk_from_str(tunnel_key->tenant_id),
+						  chunk_hash_inc(chunk_from_str(tunnel_key->account_id),
+						  chunk_hash(chunk_from_str(tunnel_key->vrf_name)))));
 }
 
 
 static void destroy_tunnel_counter_key(void *value, const void *key) {
 	(void)key;
-	struct tunnel_counter_key *p_val = (struct tunnel_counter_key*)value;
-	DESTROY_IF(p_val->super_peer_info);
+
+	struct tunnel_counter_key *tunnel_key = (struct tunnel_counter_key *)value;
+	free(tunnel_key->tenant_id);
+	free(tunnel_key->account_id);
+	free(tunnel_key->vrf_name);
 	free(value);
 }
 
-static void write_terminator_tunnels_oper(struct pm_terminator_conn* const terminator_conn) {
+CALLBACK(write_terminator_tenant_account_vrf_tunnels, void,
+		 void *ctx, const struct system_enterprise_map_iterator* iter) {
```

Comment:
The whole point of the `CALLBACK()` macro is to cast the first argument to the correct pointer type. It's useless to get it as a void* and then manually cast it.
Instead the `ctx` should point to `hashtable_t` directly.